Programmable snapshots API

766C6164 opened the pull request:
This PR relates to https://github.com/AntelopeIO/leap/issues/656 issue and introduces new feature - snapshot scheduling API to create, retrieve and delete/modify automated snapshot schedule.

In more details API specification with examples of it's use can be found here:
https://github.com/eosnetworkfoundation/product/blob/main/api-http/proposals/snapshot-api.md

This draft PR includes initial boost tests and will incorporate additional tests currently being finalized.

heifner commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp:
```diff
@@ -0,0 +1,172 @@
+   std::function<void(producer_plugin::next_function<producer_plugin::snapshot_information>)> _create_snapshot;
+
+public:
+   snapshot_scheduler() {}
```
`= default;`

766C6164 replied:
done

heifner commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp:
```diff
@@ -0,0 +1,172 @@
+         // remove expired request
+         if(req.end_block_num > 0 && height >= req.end_block_num) {
+            auto& snapshot_by_id = _snapshot_requests.get<by_snapshot_id>();
+            _snapshot_requests.erase(snapshot_by_id.find(req.snapshot_request_id));
```
I think this can just be `_snapshot_requests.erase(req.snapshot_request_id);`

766C6164 replied:
yes, that is better :) done

heifner commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp:
```diff
@@ -0,0 +1,172 @@
+               edump((e.to_detail_string()));
+               throw;
+            } catch(const std::exception& e) {
+               edump((e.what()));
```
Avoid using `edump` in production code. It is for quick debugging, not for anything committed to the code base.

766C6164 replied:
done

heifner commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp:
```diff
@@ -0,0 +1,172 @@
+         }
+      };
+      _create_snapshot(next);
+      //auto theasync=std::async([&,next]{ return _create_snapshot(next);});
```
remove

766C6164 replied:
done

heifner commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp:
```diff
@@ -0,0 +1,172 @@
+            // success
+         }
+      };
+      _create_snapshot(next);
```
Would like to see the `create_snapshot` code moved into this class and out of `producer_plugin`.

heifner commented on plugins/producer_plugin/producer_plugin.cpp:
```diff
@@ -988,6 +999,9 @@ void producer_plugin::plugin_initialize(const boost::program_options::variables_
    }
 
+   my->_snapshot_scheduler.set_db_path(my->_snapshots_dir);
+   my->_snapshot_scheduler.set_create_snapshot_fn(std::bind(&producer_plugin::create_snapshot, this, std::placeholders::_1));
```
Use a lambda instead of `std::bind`.

766C6164 replied:
done

heifner commented on plugins/producer_plugin/producer_plugin.cpp:
```diff
@@ -1013,6 +1027,7 @@ void producer_plugin::plugin_startup()
    my->_accepted_block_connection.emplace(chain.accepted_block.connect( [this]( const auto& bsp ){ my->on_block( bsp ); } ));
    my->_accepted_block_header_connection.emplace(chain.accepted_block_header.connect( [this]( const auto& bsp ){ my->on_block_header( bsp ); } ));
    my->_irreversible_block_connection.emplace(chain.irreversible_block.connect( [this]( const auto& bsp ){ my->on_irreversible_block( bsp->block ); } ));
+   my->_block_start_connection.emplace(chain.block_start.connect( [this]( uint32_t bs ){ my->_snapshot_scheduler.on_block(bs); } ));
```
Please use `on_start_block` for `block_start` signal to avoid confusion. We use `on_block` for `accepted_block` signal. Although, I think you want to attach to `accepted_block` not `block_start`.

766C6164 replied:
done

heifner commented on plugins/producer_plugin/producer_plugin.cpp:
```diff
@@ -382,6 +390,8 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
          auto before = _unapplied_transactions.size();
          _unapplied_transactions.clear_applied( bsp );
          _subjective_billing.on_block( _log, bsp, fc::time_point::now() );
+         //_snapshot_scheduler.on_block(bsp->block_num);
```
Remove. Although, I wonder why you are calling snapshot_scheduler on blocks start instead of on accepted block.

heifner replied:
It would be nice to move all the existing snapshot generation code out of producer_plugin into the snapshot_scheduler or related classes.

766C6164 replied:
on_start_block was the only hook that allowed execution of existing implementation of create_snapshot without massive chain of hard to trace exceptions. on_start_block hook executes create+snapshot perfectly with no issues.

heifner commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp:
```diff
@@ -0,0 +1,172 @@
+         // execute "asap" or if matches spacing
+         if((req.start_block_num == 0) ||
+            (!((height - req.start_block_num) % req.block_spacing))) {
+            execute_snapshot();
```
Since `on_block` is called from start block, doesn't this create a snapshot for the previous block and not the current block?

766C6164 replied:
bug, that's what will happen, fixed

heifner replied:
Would be good to add a test that verifies the correct block height snapshot is created.

766C6164 replied:
added, reading this info from pending snapshot for simplicity

heifner commented on plugins/producer_plugin/producer_plugin.cpp:
```diff
@@ -25,7 +26,10 @@
+#include <boost/multi_index/random_access_index.hpp>
 #include <boost/signals2/connection.hpp>
+#include <boost/range/adaptor/map.hpp>
+#include <boost/range/algorithm/copy.hpp>
```
None of these new boost includes are needed.

766C6164 replied:
removed

766C6164 commented on plugins/producer_plugin/include/eosio/producer_plugin/producer_plugin.hpp:
```diff
@@ -50,6 +50,19 @@ class producer_plugin : public appbase::plugin<producer_plugin> {
    };
 
+   struct snapshot_request_information {
+      uint32_t             snapshot_request_id = 0;
```
Yes, that would be definitely cleaner, thank you. Done.

766C6164 commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp:
```diff
@@ -0,0 +1,172 @@
+      }
+
+      auto request_id = sri.snapshot_request_id ? sri.snapshot_request_id : _snapshot_id++;
+      _snapshot_requests.emplace(producer_plugin::snapshot_request_information{request_id, sri.block_spacing, sri.start_block_num, sri.end_block_num, {}});
```
done, snapshot_request_id is out from request and will be always auto generated

766C6164 commented on plugins/producer_plugin/include/eosio/producer_plugin/producer_plugin.hpp:
```diff
@@ -105,7 +118,11 @@ class producer_plugin : public appbase::plugin<producer_plugin> {
+
    void create_snapshot(next_function<snapshot_information> next);
+   void schedule_snapshot(const snapshot_request_information& schedule);
+   void unschedule_snapshot(const snapshot_request_information& schedule);
```
for un-schedule replaced with struct holging just uint32_t. It arrives from http request parser and this seems a right way to do it. Then struct snapshot_schedule_information : public snapshot_request_id_information, public snapshot_request_information inherited from both which is also seems cleaner

766C6164 commented on plugins/producer_plugin/include/eosio/producer_plugin/producer_plugin.hpp:
```diff
@@ -148,7 +165,7 @@ class producer_plugin : public appbase::plugin<producer_plugin> {
 
  private:
-   std::shared_ptr<class producer_plugin_impl> my;
+   std::shared_ptr<class producer_plugin_impl>  my;
```
done

766C6164 commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_db_json.hpp:
```diff
@@ -0,0 +1,77 @@
+
+namespace eosio {
+
+using namespace eosio::chain;
```
done, removed

766C6164 commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_db_json.hpp:
```diff
@@ -0,0 +1,77 @@
+
+class snapshot_db_json {
+public:
+   snapshot_db_json() {}
```
done

766C6164 commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_db_json.hpp:
```diff
@@ -0,0 +1,77 @@
+class snapshot_db_json {
+public:
+   snapshot_db_json() {}
+   snapshot_db_json(bfs::path db_path) : db_path(db_path) {}
```
removed this ctor as it ends up never used

766C6164 commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_db_json.hpp:
```diff
@@ -0,0 +1,77 @@
+public:
+   snapshot_db_json() {}
+   snapshot_db_json(bfs::path db_path) : db_path(db_path) {}
+   ~snapshot_db_json() {}
```
done

766C6164 commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_db_json.hpp:
```diff
@@ -0,0 +1,77 @@
+   ~snapshot_db_json() {}
+
+   void set_path(bfs::path path) {
+      db_path = path;
```
done

766C6164 commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_db_json.hpp:
```diff
@@ -0,0 +1,77 @@
+
+using namespace eosio::chain;
+
+class snapshot_db_json {
```
done

766C6164 commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_db_json.hpp:
```diff
@@ -0,0 +1,77 @@
+      root.push_back(std::make_pair("snapshot_requests", node_srs));
+
+      std::ofstream file(get_json_path().string());
+      boost::property_tree::write_json(file, root);
```
done

heifner replied:
`std::ifstream` and `std::ofstream` do not throw exceptions by default. Need to enable exceptions.
`file.exceptions(std::istream::failbit|std::istream::eofbit);`

766C6164 replied:
Thank you! Done.

766C6164 commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp:
```diff
@@ -0,0 +1,172 @@
+#include <boost/multi_index/random_access_index.hpp>
+#include <boost/multi_index_container.hpp>
+
+#include <eosio/producer_plugin/snapshot_db_json.hpp>
```
done

heifner replied:
I would have made it the first include, but not a huge deal.

766C6164 commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp:
```diff
@@ -0,0 +1,172 @@
+using chain::account_name;
+using chain::block_state_ptr;
+using chain::packed_transaction;
+using chain::transaction_id_type;
```
done

heifner commented:
Would like see an integration test for this functionality.

heifner commented on plugins/producer_plugin/include/eosio/producer_plugin/producer_plugin.hpp:
```diff
@@ -158,6 +182,9 @@ FC_REFLECT(eosio::producer_plugin::greylist_params, (accounts));
 FC_REFLECT(eosio::producer_plugin::snapshot_information, (head_block_id)(head_block_num)(head_block_time)(version)(snapshot_name))
+FC_REFLECT(eosio::producer_plugin::snapshot_request_information, (block_spacing)(start_block_num)(end_block_num)(snapshot_description))
+FC_REFLECT(eosio::producer_plugin::snapshot_request_id_information, (snapshot_request_id))
+FC_REFLECT(eosio::producer_plugin::get_snapshot_requests_result, (requests))
```
I thought you had to use FC_REFLECT_DERIVED for these derived types.

766C6164 replied:
done

heifner commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_db_json.hpp:
```diff
@@ -0,0 +1,90 @@
+      }
+      catch (std::ifstream::failure e) {
+          wlog( "unable to restore snapshots schedule from filesystem: ${details}",
+                  ("details",e.what()) );
```
Should be 3 space indent, an `elog`, and indicate the path in the log.

766C6164 replied:
done

heifner commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_db_json.hpp:
```diff
@@ -0,0 +1,90 @@
+      }
+      catch (std::ofstream::failure e) {
+          wlog( "unable to store snapshots schedule to filesystem: ${details}",
+                  ("details",e.what()) );
```
Should be 3 space indent, an elog, and indicate the path in the log.

766C6164 replied:
done

heifner commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp:
```diff
@@ -0,0 +1,159 @@
+
+   // initialize with storage
+   void set_db_path(bfs::path db_path) {
+      _snapshot_db.set_path(db_path);
```
`std::move(db_path)`

766C6164 replied:
done

heifner commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp:
```diff
@@ -0,0 +1,159 @@
+   }
+   void on_block(uint32_t height) {}
+   void on_irreversible_block(uint32_t height) {}
+   void on_abort_block() {}
```
Why do we have these?

766C6164 replied:
dropped

766C6164 commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp:
```diff
@@ -0,0 +1,159 @@
+               wlog( "snapshot creation error: ${details}", ("details",e.to_detail_string()) );
+               throw;
+            } catch(const std::exception& e) {
+               wlog( "snapshot creation error: ${details}", ("details",e.what()) );
```
done

766C6164 commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp:
```diff
@@ -0,0 +1,159 @@
+            try {
+               std::get<fc::exception_ptr>(result)->dynamic_rethrow_exception();
+            } catch(const fc::exception& e) {
+               wlog( "snapshot creation error: ${details}", ("details",e.to_detail_string()) );
```
done

766C6164 commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp:
```diff
@@ -0,0 +1,159 @@
+
+   // snapshot executor
+   void set_create_snapshot_fn(std::function<void(producer_plugin::next_function<producer_plugin::snapshot_information>)> fn) {
+      _create_snapshot = fn;
```
done

766C6164 commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp:
```diff
@@ -0,0 +1,159 @@
+   virtual producer_plugin::get_snapshot_requests_result get_snapshot_requests() {
+      producer_plugin::get_snapshot_requests_result result;
+      auto& asvector = _snapshot_requests.get<as_vector>();
+      result.requests.insert(result.requests.begin(), asvector.begin(), asvector.end());
```
done

greg7mdp replied:
Actually the `reserve()` is not needed because insert() already does it if needed.

heifner commented on tests/nodeos_snapshot_programmable_test.py:
```diff
@@ -0,0 +1,210 @@
+#!/usr/bin/env python3
+
```
This is almost identical to `nodeos_snapshot_diff_test.py`. Seems like they could easily be combined into one test that does both.

766C6164 replied:
Thank you! I updated nodeos_snapshot_diff_test to use scheduler and then compare that snapshot to irreversible.  I am working on re-purposing nodeos_snapshot_programmable_test for fork scenario

heifner replied:
Are you planning on removing this test then?

greg7mdp commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp:
```diff
@@ -0,0 +1,189 @@
+   struct by_snapshot_value;
+   struct as_vector;
+
+   using snapshot_requests = bmi::multi_index_container<
+         producer_plugin::snapshot_schedule_information,
+         indexed_by<
+               bmi::hashed_unique<tag<by_snapshot_id>, BOOST_MULTI_INDEX_MEMBER(producer_plugin::snapshot_request_id_information, uint32_t, snapshot_request_id)>,
+               bmi::random_access<tag<as_vector>>,
+               bmi::ordered_unique<tag<by_snapshot_value>,
+                                   composite_key<producer_plugin::snapshot_schedule_information,
+                                                 BOOST_MULTI_INDEX_MEMBER(producer_plugin::snapshot_request_information, uint32_t, block_spacing),
+                                                 BOOST_MULTI_INDEX_MEMBER(producer_plugin::snapshot_request_information, uint32_t, start_block_num),
+                                                 BOOST_MULTI_INDEX_MEMBER(producer_plugin::snapshot_request_information, uint32_t, end_block_num)>>>>;
```
I feel that using a `multi-index-container`  is overkill here, and adds unnecessary complexity. I would suggest using an `unordered_set<producer_plugin::snapshot_schedule_information>` with the hash and compare done on the `snapshot_request_id`.

The serialization could as easily be done from the set as from the `<as_vector>` view.
The `<by_snapshot_value>` is used only to find out if a new request is a duplicate, and I think that doing a linear search is probably OK since we are unlikely to have many thousands of snapshot requests.

After all we already do a linear search at every `on_start_block`.

greg7mdp replied:
@Vlad, I understand that you had a requirement to use `multi_index_container`, and not time to revisit this choice for now, so marking this as resolved.  Thanks!

766C6164 commented on plugins/producer_plugin/include/eosio/producer_plugin/producer_plugin.hpp:
```diff
@@ -51,6 +51,26 @@ class producer_plugin : public appbase::plugin<producer_plugin> {
+   };
+
+   struct snapshot_request_id_information {
+      uint32_t                            snapshot_request_id = 0;
```
Thank you! Leftover after removing other long-typed struct members. Fixed.

greg7mdp commented on plugins/producer_plugin/test/test_snapshot_scheduler.cpp:
```diff
@@ -0,0 +1,140 @@
+
+BOOST_AUTO_TEST_SUITE(snapshot_scheduler_test)
+
+BOOST_AUTO_TEST_CASE(snapshot_scheduler_test) {
```
Nice test!

heifner commented on tests/nodeos_snapshot_diff_test.py:
```diff
@@ -152,10 +156,9 @@ def waitForBlock(node, blockNum, blockType=BlockType.head, timeout=None, reportI
 
     Print("Create snapshot")
-    ret = nodeSnap.createSnapshot()
+    ret = nodeSnap.scheduleSnapshot()
```
I think this is probably fine. Would have been better to enhance this test to do both, but not a huge deal.

766C6164 replied:
Done, now 3 snapshots are being generated and compared to each other

heifner commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_db_json.hpp:
```diff
@@ -0,0 +1,90 @@
+      }
+      catch (std::ifstream::failure & e) {
+         elog( "unable to restore snapshots schedule from filesystem ${jsonpath}, details: ${details}",
+            ("jsonpath", get_json_path().string()) ("details",e.what()) );
```
I think we should consider an `app().quit()` here and below. Typically users will dedicate a node for creating snapshots. If this node is unable to do so, seems like it should indicate an error and exit.

766C6164 replied:
good point, agreed, done. Thank you!

heifner commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp:
```diff
@@ -0,0 +1,189 @@
+               throw;
+            } catch(const std::exception& e) {
+               elog( "snapshot creation error: ${details}", ("details",e.what()) );
+               throw;
```
Does this result in a graceful shutdown? Can you manually verify that it does.

766C6164 replied:
No, it was abort, changed this to app().quit() and now shutdown graceful. Thank you!!!

766C6164 commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp:
```diff
@@ -0,0 +1,189 @@
+   snapshot_scheduler() = default;
+
+   // snapshot_scheduler_listener
+   void on_start_block(uint32_t height) {
+      for(const auto& req: _snapshot_requests.get<0>()) {
+         // assume "asap" for snapshot with missed/zero start, it can have spacing
+         if(!req.start_block_num) {
+            // update start_block_num with current height only if this is recurring
+            // if non recurring, will be executed and unscheduled
+            if (req.block_spacing && height) {
+               auto& snapshot_by_id = _snapshot_requests.get<by_snapshot_id>();
+               auto it = snapshot_by_id.find(req.snapshot_request_id);
+               _snapshot_requests.modify(it, [&height](auto& p) { p.start_block_num = height; });
+               x_serialize();
+            }
+            execute_snapshot(req.snapshot_request_id);
+         }
+         // execute snapshot, -1 since its called from start block
+         else if(!req.block_spacing || (!((height - req.start_block_num - 1) % req.block_spacing))) {
+            execute_snapshot(req.snapshot_request_id);
+         }       
+
+         // cleanup - remove expired (or invalid) request
+         if(!req.start_block_num || !req.block_spacing || (req.end_block_num > 0 && height >= req.end_block_num)) {
+            unschedule_snapshot(req.snapshot_request_id);
+         }
+      }
+   }
```
Great suggestion, done

766C6164 commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp:
```diff
@@ -0,0 +1,189 @@
+
+   // initialize with storage
+   void set_db_path(bfs::path db_path) {
+      _snapshot_db.set_path(std::move(db_path));
```
yes, or probably clearing it up before insert is what should be expected behavior. done. Thank you!

766C6164 commented on plugins/producer_plugin/test/test_snapshot_scheduler.cpp:
```diff
@@ -0,0 +1,140 @@
+         producer_plugin::snapshot_request_information sri2 = {.block_spacing = 5000, .start_block_num = 100000, .end_block_num = 300000, .snapshot_description = "Example of recurring snapshot 2"};
+         producer_plugin::snapshot_request_information sri3 = {.block_spacing = 2, .start_block_num = 0, .end_block_num = 3, .snapshot_description = "Example of recurring snapshot 1"};
+
+         pp->schedule_snapshot(sri1);
```
This is tricky, I am catching firing of sri1 snapshot request, which will immediately becomes pending and happening on start=1 + spacing=8. I'll see if I can do it in more clear/deterministic way. Thank you!

766C6164 commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp:
```diff
@@ -0,0 +1,172 @@
+   virtual void schedule_snapshot(const snapshot_request_information& sri) = 0;
+   virtual void unschedule_snapshot(uint32_t sri) = 0;
+   virtual get_snapshot_requests_result get_snapshot_requests() = 0;
+};
```
Dropped these, thank you!

greg7mdp commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp:
```diff
@@ -0,0 +1,186 @@
+            execute_snapshot(req.snapshot_request_id);
+         }
+         // execute snapshot, -1 since its called from start block
+         else if(!req.block_spacing || (!((height - req.start_block_num - 1) % req.block_spacing))) {
```
I believe this test is incorrect. Even if the snapshot specifies a block far in the future (for example `{.block_spacing = 0, .start_block_num = 1500, .end_block_num = 1500`, the snapshot will be taken the first time `on_start_block` is called after the snapshot is defined.

766C6164 replied:
You totally right, i was tweaking this code and accidentally broke it. I tried to make it more readable now. Thank you!!!!

greg7mdp commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp:
```diff
@@ -0,0 +1,186 @@
+
+         // cleanup - remove expired (or invalid) request
+         if(!req.start_block_num || !req.block_spacing || (req.end_block_num > 0 && height >= req.end_block_num)) {
+            unschedule_snapshot(req.snapshot_request_id);
```
suggestion: add `serialize_needed = true;` here and remove it from `unschedule_snapshot()`

766C6164 replied:
unschedule_snapshot is also called directly from producer plugin that  expects serialization to happen as well

greg7mdp replied:
OK, I missed that. So we could leave it as is, or add a parameter `bool serialize` to `unschedule_snapshot`. Up to you!

greg7mdp commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp:
```diff
@@ -0,0 +1,186 @@
+               _snapshot_requests.modify(it, [&height](auto& p) { p.start_block_num = height; });
+               serialize_needed = true;              
+            }
+            execute_snapshot(req.snapshot_request_id);
```
Whenever we call `execute_snapshot()`, we should `break;` out of the loop, as there is no reason to execute two snapshots at the same block height.

766C6164 replied:
Agreed, done. Thank you!

greg7mdp commented:
We really should have a debug log trace when we are taking snapshots I think.

greg7mdp commented:
If `block_spacing == 0`, should we require that `start_block == end_block`?
If they are different, are we taking the snapshot at `start_block` or `end_block`?

greg7mdp commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp:
```diff
@@ -0,0 +1,186 @@
+   }
+
+   void execute_snapshot(uint32_t srid) {
+      _inflight_sid = srid;
```
I think we should have a `dlog` call here indicating that we are taking a snapshot, and including the snapshot_request parameters (`start_block_num`, `end_block_num`, `block_spacing`), as well as the current block height.

766C6164 replied:
Thank you! Planned to do it, but forgot. Added dlog.

greg7mdp replied:
Thanks for adding it!

766C6164 commented:
> If `block_spacing == 0`, should we require that `start_block == end_block`? If they are different, are we taking the snapshot at `start_block` or `end_block`?

by default start_block is used per our api specification and end_block ignored. In most cases user wont specify end_block at all (by default 0) and validation for end > start wont happen in this case. I am open to changing that if that sounds wrong

greg7mdp commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp:
```diff
@@ -0,0 +1,205 @@
+      bool serialize_needed  = false;
+      bool snapshot_executed = false; 
+
+      auto execute_snapshot_with_log = [this, &height](const auto & req) {
+         dlog("snapshot scheduler creating a snapshot from the request [start_block_num:${start_block_num}, end_block_num=${end_block_num}, block_spacing=${block_spacing}], height=${height}",
+             ("start_block_num", req.start_block_num)
+             ("end_block_num",   req.end_block_num)
+             ("block_spacing",   req.block_spacing)
+             ("height",          height));
+             
+         execute_snapshot(req.snapshot_request_id);
+      };    
+
+      for(const auto& req: _snapshot_requests.get<0>()) {
+         // -1 since its called from start block
+         bool recurring_snapshot =  req.block_spacing &&  (height >= req.start_block_num + 1) && (!((height - req.start_block_num - 1) % req.block_spacing));
+         bool onetime_snapshot   = (!req.block_spacing) && (height == req.start_block_num + 1);
+        
+         // assume "asap" for snapshot with missed/zero start, it can have spacing
+         if(!req.start_block_num) {
+            // update start_block_num with current height only if this is recurring
+            // if non recurring, will be executed and unscheduled
+            if (req.block_spacing && height) {
+               auto& snapshot_by_id = _snapshot_requests.get<by_snapshot_id>();
+               auto it = snapshot_by_id.find(req.snapshot_request_id);
+               _snapshot_requests.modify(it, [&height](auto& p) { p.start_block_num = height; });
+               serialize_needed = true;              
+            }
+            execute_snapshot_with_log(req);
+            snapshot_executed = true;
+         }        
+         else if(recurring_snapshot || onetime_snapshot) {
+            execute_snapshot_with_log(req);
+         }       
+
+         // cleanup - remove expired (or invalid) request
+         if(!req.start_block_num || !req.block_spacing || (req.end_block_num > 0 && height >= req.end_block_num)) {
+            unschedule_snapshot(req.snapshot_request_id);
+            snapshot_executed = true;
+         }
```
```suggestion
      bool snapshot_executed = false; 

      auto execute_snapshot_with_log = [this, &height, &snapshot_executed](const auto & req) {
         dlog("snapshot scheduler creating a snapshot from the request [start_block_num:${start_block_num}, end_block_num=${end_block_num}, block_spacing=${block_spacing}], height=${height}",
             ("start_block_num", req.start_block_num)
             ("end_block_num",   req.end_block_num)
             ("block_spacing",   req.block_spacing)
             ("height",          height));
             
         execute_snapshot(req.snapshot_request_id);
         snapshot_executed = true;
      };    

      for(const auto& req: _snapshot_requests.get<0>()) {
         // -1 since its called from start block
         bool recurring_snapshot =  req.block_spacing &&  (height >= req.start_block_num + 1) && (!((height - req.start_block_num - 1) % req.block_spacing));
         bool onetime_snapshot   = (!req.block_spacing) && (height == req.start_block_num + 1);
        
         // assume "asap" for snapshot with missed/zero start, it can have spacing
         if(!req.start_block_num) {
            // update start_block_num with current height only if this is recurring
            // if non recurring, will be executed and unscheduled
            if (req.block_spacing && height) {
               auto& snapshot_by_id = _snapshot_requests.get<by_snapshot_id>();
               auto it = snapshot_by_id.find(req.snapshot_request_id);
               _snapshot_requests.modify(it, [&height](auto& p) { p.start_block_num = height; });
               serialize_needed = true;              
            }
            execute_snapshot_with_log(req);
         }        
         else if(recurring_snapshot || onetime_snapshot) {
            execute_snapshot_with_log(req);
         }       

         // cleanup - remove expired (or invalid) request
         if(!req.start_block_num || !req.block_spacing || (req.end_block_num > 0 && height >= req.end_block_num)) {
            unschedule_snapshot(req.snapshot_request_id);
         }
```

greg7mdp commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp:
```diff
@@ -0,0 +1,205 @@
+         // cleanup - remove expired (or invalid) request
+         if(!req.start_block_num || !req.block_spacing || (req.end_block_num > 0 && height >= req.end_block_num)) {
+            unschedule_snapshot(req.snapshot_request_id);
+            snapshot_executed = true;
```
this `snapshot_executed = true;` should be after line 88. Even better capture `&snapshot_executed` in the `execute_snapshot_with_log` lambda and set it to true there.

766C6164 replied:
Ups... thank you!!! Fixed

greg7mdp commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp:
```diff
@@ -0,0 +1,205 @@
+         }       
+
+         // cleanup - remove expired (or invalid) request
+         if(!req.start_block_num || !req.block_spacing || (req.end_block_num > 0 && height >= req.end_block_num)) {
```
Hum, I believe this would remove all requests with `!req.block_spacing ` even before they are executed.

766C6164 replied:
Right! Thank you, appreciate it! I adjusted handling of this case and tested this scenario.

greg7mdp commented:
What about this simpler version:

```
   void on_start_block(uint32_t height) {
      bool serialize_needed  = false;

      for(const auto& req: _snapshot_requests.get<0>()) {
         if(!req.start_block_num && height) {
            // update start_block_num with current height only if not set
            auto& snapshot_by_id = _snapshot_requests.get<by_snapshot_id>();
            auto it = snapshot_by_id.find(req.snapshot_request_id);
            _snapshot_requests.modify(it, [&height](auto& p) { p.start_block_num = height - 1; });
            serialize_needed = true;              
         }
      }
      
      // store db to filesystem
      if (serialize_needed)
         x_serialize();

      for(const auto& req: _snapshot_requests.get<0>()) {   
         // -1 since its called from start block
         bool recurring_snapshot =  req.block_spacing &&  (height >= req.start_block_num + 1) && (!((height - req.start_block_num - 1) % req.block_spacing));
         bool onetime_snapshot   = (!req.block_spacing) && (height == req.start_block_num + 1);
        
         if(recurring_snapshot || onetime_snapshot) {
            dlog("snapshot scheduler creating a snapshot from the request [start_block_num:${start_block_num}, end_block_num=${end_block_num}, block_spacing=${block_spacing}], height=${height}",
                 ("start_block_num", req.start_block_num)
                 ("end_block_num",   req.end_block_num)
                 ("block_spacing",   req.block_spacing)
                 ("height",          height));
             
            execute_snapshot(req.snapshot_request_id);
            if (onetime_snapshot)
               unschedule_snapshot(req.snapshot_request_id);
            break; // stop iterating snapshot requests after snapshot execution
         }

         // cleanup - remove expired (or invalid) request
         if(req.block_spacing && (req.end_block_num > 0 && height >= req.end_block_num))
            unschedule_snapshot(req.snapshot_request_id);
      }
   }
```

766C6164 commented:
> What about this simpler version (untested):
> 
> ```
>    void on_start_block(uint32_t height) {
>       bool serialize_needed  = false;
> 
>       for(const auto& req: _snapshot_requests.get<0>()) {
>          if(!req.start_block_num && height) {
>             // update start_block_num with current height only if not set
>             auto& snapshot_by_id = _snapshot_requests.get<by_snapshot_id>();
>             auto it = snapshot_by_id.find(req.snapshot_request_id);
>             _snapshot_requests.modify(it, [&height](auto& p) { p.start_block_num = height - 1; });
>             serialize_needed = true;              
>          }
>       }
>       
>       // store db to filesystem
>       if (serialize_needed)
>          x_serialize();
> 
>       for(const auto& req: _snapshot_requests.get<0>()) {   
>          // -1 since its called from start block
>          bool recurring_snapshot =  req.block_spacing &&  (height >= req.start_block_num + 1) && (!((height - req.start_block_num - 1) % req.block_spacing));
>          bool onetime_snapshot   = (!req.block_spacing) && (height == req.start_block_num + 1);
>         
>          if(recurring_snapshot || onetime_snapshot) {
>             dlog("snapshot scheduler creating a snapshot from the request [start_block_num:${start_block_num}, end_block_num=${end_block_num}, block_spacing=${block_spacing}], height=${height}",
>                  ("start_block_num", req.start_block_num)
>                  ("end_block_num",   req.end_block_num)
>                  ("block_spacing",   req.block_spacing)
>                  ("height",          height));
>              
>             execute_snapshot(req.snapshot_request_id);
>             if (onetime_snapshot)
>                unschedule_snapshot(req.snapshot_request_id);
>             break; // stop iterating snapshot requests after snapshot execution
>          }
> 
>          // cleanup - remove expired (or invalid) request
>          if(req.block_spacing && req.end_block_num > 0 && height >= req.end_block_num)
>             unschedule_snapshot(req.snapshot_request_id);
>       }
>    }
> ```

That looks good, thank you! Before I saw this I adjusted my code to iterate through requests once, but completely since in one pass we are correcting missing start_block_num's and also executing snapshot if needed, but only once. Therefore I removed break, replaced with if in lambda

greg7mdp commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp:
```diff
@@ -0,0 +1,206 @@
+            unschedule_snapshot(req.snapshot_request_id);
+         }
+      }
+
```
There should be `if (snapshot_executed) break;` here.

766C6164 replied:
Yes, makes sense, added. Thank you!

greg7mdp commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp:
```diff
@@ -0,0 +1,207 @@
+            (!req.block_spacing && height >= (req.start_block_num + 1)) || 
+            (req.end_block_num > 0 && height >= (req.end_block_num + 1))) {
+            unschedule_snapshot(req.snapshot_request_id);
+            if (snapshot_executed) break;
```
This `if` statement should be one line further (after the closing brace).

766C6164 replied:
This way it will break loop that may have another start_block_num's to correct... It needs to be completed fully and block for execution of snapshot is now in execute_snapshot_with_log lambda. What do you think?

greg7mdp replied:
Yes, but we don't want to take two snapshots at the same block height.
In any case, if we already too a snapshot, it is OK to update the `start_block_num` the next time this function gets called (at height+1).

greg7mdp replied:
Oh, yes, right, I missed that. So we probably can remove line 99, right?

766C6164 replied:
done! thanks again!

greg7mdp commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp:
```diff
@@ -0,0 +1,207 @@
+         }       
+
+         // cleanup - remove expired (or invalid) request
+         if(!req.start_block_num ||         
```
This check `(!req.start_block_num || ` will remove all recurring snapshot requests which have `start_block_num == 0`, even if they are not expired. My version with two loops fixed that.

766C6164 replied:
I am pretty sure at this line req.start_block_num==0 should not ever happen, being corrected few lines above, but I agree, your suggestion is safest one! Thank you!

greg7mdp replied:
> I am pretty sure at this line req.start_block_num==0 should not ever happen, being corrected few lines above, but I agree, your suggestion is safest one! Thank you!

I wasn't sure, because I see we update the request in boost multi-index, but `req` is a `const &`  to a view of that multi-index and I was not sure it was necessarily updated.

greg7mdp commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp:
```diff
@@ -0,0 +1,207 @@
+         }       
+
+         // cleanup - remove expired (or invalid) request
+         if(!req.start_block_num ||         
```
```suggestion
         if((!req.start_block_num && !req.block_spacing) ||         
```

766C6164 replied:
Done, thank you!!!

greg7mdp commented:
@766C6164 Thanks for all the changes! I think this is going to be a very popular feature :-).

Lin merged his large PR to main a few minutes ago. Maybe you should merge the latest changes from main (hopefully without too main conflicts), and then I'll do a last check and approve your PR, probably tomorrow morning.

766C6164 commented:
> @766C6164 Thanks for all the changes! I think this is going to be a very popular feature :-).
> 
> Lin merged his large PR to main a few minutes ago. Maybe you should merge the latest changes from main (hopefully without too main conflicts), and then I'll do a last check and approve your PR, probably tomorrow morning.

Thank you so much!!! Really appreciate all your time and effort you put into reviewing this PR! Luckily picked up changes from main with only 3 minor conflicts :)

linh2931 commented on plugins/producer_api_plugin/producer_api_plugin.cpp:
```diff
@@ -104,6 +109,18 @@ void producer_api_plugin::plugin_startup() {
+       CALL_WITH_400(producer, producer, get_integrity_hash,
+            INVOKE_R_V(producer, get_integrity_hash), 201),
+       CALL_ASYNC(producer, producer, create_snapshot, producer_plugin::snapshot_information,
+            INVOKE_R_V_ASYNC(producer, create_snapshot), 201),
```
Merge conflicts: `set_whitelist_blacklist`, `get_integrity_hash`, `get_integrity_hash` should be removed from here. They are not safe to read-only transactions and defined from line 147. Sorry for the trouble my PR causes.

linh2931 replied:
I checked out your branch and did a test. Confirmed the about problem caused CICD failures:

error 2023-03-16T13:52:42.400 nodeos    main.cpp:161                  main                 ] 3110006  plugin_config_exception: Incorrect plugin configuration
http url /v1/producer/create_snapshot is not unique
    {"u":"/v1/producer/create_snapshot"}
    nodeos  http_plugin.cpp:370 add_handler

linh2931 commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp:
```diff
@@ -0,0 +1,206 @@
+                     p.pending_snapshots = std::move(pending);
+                  });
+               }
+            }
```
Do you want to report an error if `snapshot_req` was not found?

766C6164 replied:
done

766C6164 replied:
not sure if we should gracefully exit here, probably not. un-finalized snapshot (should not happen) does not sound like terminal case.

linh2931 commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp:
```diff
@@ -0,0 +1,206 @@
+            }
+         }
+      };
+      _create_snapshot(next);
```
https://github.com/AntelopeIO/leap/issues/830

heifner replied:
Would be nice for that to be fixed before this is merged if possible, as that could cause a segfault I believe.

766C6164 replied:
so @linh2931 do you think we should not change this? Thank you!

greg7mdp commented on plugins/producer_plugin/include/eosio/producer_plugin/producer_plugin.hpp:
```diff
@@ -82,6 +82,26 @@ class producer_plugin : public appbase::plugin<producer_plugin> {
+   };
+
+   struct snapshot_schedule_information : public snapshot_request_id_information, public snapshot_request_information {
+      std::optional<std::vector<snapshot_information>> pending_snapshots;
```
I don't see the benefit of making this a `std::optional`. Isn't just an empty vector the same?

Also, Just wondering, is this pending_snapshots vector, and its management in the functions  add_pending_snapshot_info and execute_snapshot used only in the tests?

766C6164 replied:
no, this is exposed as api call, snapshot can take significant time to finalize so if there are any pending snapshot(s) triggered by a snapshot request, it will be returned in reply as json showing that it's in progress...

766C6164 commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_db_json.hpp:
```diff
@@ -0,0 +1,92 @@
+         std::ifstream file(get_json_path().string());
+         file.exceptions(std::istream::failbit|std::istream::eofbit);
+         boost::property_tree::read_json(file, root);
+         file.close();
```
done, thank you!

766C6164 commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_db_json.hpp:
```diff
@@ -0,0 +1,92 @@
+         std::ofstream file(get_json_path().string());
+         file.exceptions(std::istream::failbit|std::istream::eofbit);
+         boost::property_tree::write_json(file, root);
+         file.close();
```
done!

766C6164 commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_db_json.hpp:
```diff
@@ -0,0 +1,92 @@
+      return *this;
+   }
+
+   const snapshot_db_json& operator<<(std::vector<producer_plugin::snapshot_schedule_information>& sr) {
```
done

766C6164 commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp:
```diff
@@ -0,0 +1,206 @@
+      bool serialize_needed  = false;
+      bool snapshot_executed = false; 
+
+      auto execute_snapshot_with_log = [this, &height, &snapshot_executed](const auto & req) {
```
done

766C6164 commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp:
```diff
@@ -0,0 +1,206 @@
+            }
+            execute_snapshot_with_log(req);
+         }        
+         else if(recurring_snapshot || onetime_snapshot) {
```
done

greg7mdp commented on plugins/producer_plugin/test/test_snapshot_scheduler.cpp:
```diff
@@ -0,0 +1,140 @@
+         auto bs = chain_plug->chain().block_start.connect([&pp](uint32_t bn) {
+            // catching pending snapshot
+            if (pp->get_snapshot_requests().snapshot_requests.begin()->pending_snapshots && pp->get_snapshot_requests().snapshot_requests.begin()->pending_snapshots->size()==1) {
+               // lets check the head block num of it, it should be 8 + 1 = 9
+               // this means we are getting a snapshot for correct block # as well
+               BOOST_CHECK_EQUAL(9, pp->get_snapshot_requests().snapshot_requests.begin()->pending_snapshots->begin()->head_block_num);   
```
We never get here, the `pending_snapshots` vector is always empty.

greg7mdp replied:
You are right, we get there. I'm not understanding why the check doesn't fail when we take a snapshot at block 3:

`snapshot scheduler creating a snapshot from the request [start_block_num:3, end_block_num=3, block_spacing=2], height=4`

766C6164 replied:
It fails for me in this case and fails hard, but not in this check if I modify sri1. Problem with start=end request that it will be removed immediately after execution so this code checking for "9" wont be reached

greg7mdp commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp:
```diff
@@ -0,0 +1,208 @@
+               }
+            } else {     
+               // should not happen, but if happens - not a terminal error          
+               elog( "unable to finalize pending snapshot request: ${details}", ("details", srid) );
```
After your last change (addition of this line), I have the following error when running the test:

![image](https://user-images.githubusercontent.com/1496907/225733044-993714ec-f7f0-40eb-ba4d-ea5ef5205d80.png)

greg7mdp replied:
I believe we hit this because the snapshot_request has already been removed from `_snapshot_requests` by the time `add_pending_snapshot_info` is called (even though we save its id in `_inflight_sid` - and I'm not sure why we do that - it is erased from the multi-index in `unschedule_snapshot`).

I suggest you just comment out this `elog` call for now - with a comment as why it is triggered.

766C6164 replied:
Thats totally correct, I was too fast adding this trace. It's normal behavior to have snapshot request removed (expired) and therefore any pendings should be gone as well.

greg7mdp commented on plugins/producer_plugin/test/test_snapshot_scheduler.cpp:
```diff
@@ -0,0 +1,140 @@
+         auto pp = appbase::app().find_plugin<producer_plugin>();
+         auto bs = chain_plug->chain().block_start.connect([&pp](uint32_t bn) {
+            // catching pending snapshot
+            if (pp->get_snapshot_requests().snapshot_requests.begin()->pending_snapshots && pp->get_snapshot_requests().snapshot_requests.begin()->pending_snapshots->size()==1) {
+               // lets check the head block num of it, it should be 8 + 1 = 9
+               // this means we are getting a snapshot for correct block # as well
+               BOOST_CHECK_EQUAL(9, pp->get_snapshot_requests().snapshot_requests.begin()->pending_snapshots->begin()->head_block_num);   
```
```suggestion
            auto& pending = pp->get_snapshot_requests().snapshot_requests.begin()->pending_snapshots;
            if (pending && pending->size()==1) {
               // lets check the head block num of it, it should be 8 + 1 = 9
               // this means we are getting a snapshot for correct block # as well
               BOOST_CHECK_EQUAL(9, pending->begin()->head_block_num);   
```

766C6164 replied:
thats totally looks cleaner! thank you!

linh2931 commented on plugins/producer_plugin/include/eosio/producer_plugin/snapshot_scheduler.hpp:
```diff
@@ -197,8 +197,8 @@ class snapshot_scheduler {
-               // should not happen, but if happens - not a terminal error          
-               elog( "unable to finalize pending snapshot request: ${details}", ("details", srid) );
+               // this could happen when request is already "expired" so there is no value in reporting it          
+               // dlog( "unable to finalize pending snapshot request: ${details}", ("details", srid) );
```
We normally do not leave commented out code behind. We remove them.

766C6164 replied:
lets remove it! thank you!

greg7mdp commented on plugins/producer_plugin/test/test_snapshot_scheduler.cpp:
```diff
@@ -0,0 +1,141 @@
+         auto pp = appbase::app().find_plugin<producer_plugin>();
+         auto bs = chain_plug->chain().block_start.connect([&pp](uint32_t bn) {
+            // catching pending snapshot
+            auto& pending = pp->get_snapshot_requests().snapshot_requests.begin()->pending_snapshots;
+            if (pending && pending->size()==1) {
+               // lets check the head block num of it, it should be 8 + 1 = 9
+               // this means we are getting a snapshot for correct block # as well
+               BOOST_CHECK_EQUAL(9, pending->begin()->head_block_num);   
+            }
```
This could happen if block_start is called before the snapshots are scheduled, of if we schedules only one non-recurring snapshot.

```suggestion
            if (!pp->get_snapshot_requests().snapshot_requests.empty()) {
               auto& pending = pp->get_snapshot_requests().snapshot_requests.begin()->pending_snapshots;
               if (pending && pending->size()==1) {
                  // lets check the head block num of it, it should be 8 + 1 = 9
                  // this means we are getting a snapshot for correct block # as well
                  BOOST_CHECK_EQUAL(9, pending->begin()->head_block_num);   
               }
            }
```

766C6164 replied:
agreed, done
